numpy/lib/recfunctions.py

```diff
@@ -960,32 +971,33 @@ def join_by(key, r1, r2, jointype='inner', r1postfix='1', r2postfix='2',
     #
     # Build the new description of the output array .......
     # Start with the key fields
-    ndtype = [list(_) for _ in r1k.dtype.descr]
+    ndtype = [list(f) for f in get_fieldspec(r1k.dtype)]
     # Add the other fields
-    ndtype.extend(list(_) for _ in r1.dtype.descr if _[0] not in key)
-    # Find the new list of names (it may be different from r1names)
-    names = list(_[0] for _ in ndtype)
-    for desc in r2.dtype.descr:
-        desc = list(desc)
-        name = desc[0]
+    ndtype.extend(list(f) for f in get_fieldspec(r1.dtype) if f[0] not in key)
+
+    for field in get_fieldspec(r2.dtype):
+        field = list(field)
         # Have we seen the current name already ?
-        if name in names:
-            nameidx = ndtype.index(desc)
+        name = field[0]
+        names = list(_[0] for _ in ndtype)
+        try:
+            nameidx = names.index(name)
+        except ValueError:
+            #... we haven't: just add the description to the current list
+            ndtype.append(field)
+        else:
             current = ndtype[nameidx]
-            # The current field is part of the key: take the largest dtype
             if name in key:
-                current[-1] = max(desc[1], current[-1])
-            # The current field is not part of the key: add the suffixes
+                # The current field is part of the key: take the largest dtype
+                current[1] = max(field[1], current[1])
```

Comment:
This was horribly incorrect before - not only would it fail to understand a `descr` like `('a', 'i4', 2)` correctly, but it would use `max('i1', 'f8') == i1`!, rather than `max(np.int8, np.float64) == np.float64)`.

This actually still doesn't work for different dtypes, as the `concatenate` fails to cast the keys to the correct types